zim/demo-library/demo-generator.py

```diff
@@ -0,0 +1,398 @@
+    NAMES_MAP.values()
+)
+
+
+FILENAME_FMT = re.compile(
+    r"^(?P<project>.+?_)(?P<lang>[a-z\-]{2,10}?_|)"
+    r"(?P<option>.+_|)(?P<year>[\d]{4}|)\-(?P<month>[\d]{2})(?P<version>[a-z]{0,2})$",
+    re.IGNORECASE,
+)
+
+REQUESTS_TIMEOUT = 10
+
+
+# Define Pydantic models
+class Demo(BaseModel):
+    name: str
+    zims: list[str]
+    expired_on: datetime.date
+
+
+class Config(BaseModel):
+    demos: list[Demo]
+
+
+class ZimInfo(BaseModel):
+    path: Path
+    infos: dict[str, Any] = {}
+
+
+def binary_human_size(size: float):
+    return human_size(size, binary=True)
+
+
+def install_demo(demo: Demo, demo_tmpl: Template):
+    demo_path = demos_path / demo.name
+    demo_path.mkdir(exist_ok=True)
+    (demo_path / "index.html").write_text(
+        demo_tmpl.render(
+            {"zims": [zims[zim_archive].infos for zim_archive in demo.zims]}
+        )
+    )
+
+
+# constants
+demos_path = Path("/data/demo")
+normal_zim_path = Path("/data/zim")
+hidden_zim_path = Path("/data/hidden-zim")
+demo_library = demos_path / "demo_library.xml"
+demo_config = demos_path / "demos.yaml"
+demo_dead_html = demos_path / "dead.html"
+demo_favicon = demos_path / "favicon.ico"
+demo_dead_kiwix_png = demos_path / "dead-kiwix.png"
+demo_home_tmpl_path = demos_path / "demo-home.jinja2"
+favicons_path = demos_path / "favicons"
+favicons_path.mkdir(exist_ok=True)
+logger = getLogger("demo-generator")
+assets_base_url = (
+    "https://raw.githubusercontent.com/kiwix/operations/demo-library/zim/demo-library/"
+)
+logger.info(f"Assets will be downloaded from {assets_base_url}")
+varnish_cache_url = "http://demo-library-webcache-service"
+
+
+def zim_date_key(s: str) -> tuple[str, int, str]:
+    """Function to be used as sorting criteria for ZIM names.
+
+    Function properly sort by name then date, including versions suffix in the date
+
+    E.g. sorted(
+        [
+            "tests_eng_test_2024-01aa",
+            "tests_eng_test_2024-01b",
+            "tests_eng_test_2024-01",
+            "tests_eng_test_2023-01",
+        ],
+        key=zim_date_key,
+    ) == [
+        "tests_eng_test_2023-01",
+        "tests_eng_test_2024-01",
+        "tests_eng_test_2024-01b",
+        "tests_eng_test_2024-01aa",
+    ]
+
+    The "trick" is to return first up to the xxxx-xx part, then the length of optional
+    ll suffix, then the suffix itself. Python takes care of the rest (first compare the
+    prefix, then the lenght of the suffix, then the suffix itself)
+    """
+    underscore_pos = s.rfind("_")
+    dot_pos = s.rfind(".")
+    return (
+        s[: underscore_pos + 8],
+        len(s[underscore_pos + 8 : dot_pos]),
+        s[underscore_pos + 8 : dot_pos],
+    )
+
+
+class BadZimNameError(Exception):
+    """ZIM matching passed name has not been found on the storage"""
+
+    pass
+
+
+class ZimNotFoundError(Exception):
+    """ZIM has not been found on the storage"""
+
+    pass
+
+
+def get_ziminfo(zim_archive: str) -> ZimInfo:
```

Comment:
xShould be a `Path`. This implies several changes in the function. All for the best IMO

Reply:
I don't agree, as mentioned in docstring, it can be an incomplete path ; of course I could abuse the Path and it will work, but I do not see it as an added value

Reply:
If you don't see it then it's your call